Upstream `sklearn` change causes error ('super' object has no attribute '__sklearn_tags__') in test suite

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Bug description

The current test suite yields the following error

```
____________________ ERROR collecting tests/test_sklearn.py ____________________
tests/test_sklearn.py:2[56](https://github.com/AdaptiveMotorControlLab/CEBRA/actions/runs/12357735292/job/34486685645?pr=190#step:10:57): in <module>
    @_util.parametrize_with_checks_slow(
tests/_util.py:66: in parametrize_with_checks_slow
    fast_params = [
tests/_util.py:67: in <listcomp>
    list(
/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/sklearn/utils/estimator_checks.py:543: in estimator_checks_generator
    for check in _yield_all_checks(estimator, legacy=legacy):
/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/sklearn/utils/estimator_checks.py:3[59](https://github.com/AdaptiveMotorControlLab/CEBRA/actions/runs/12357735292/job/34486685645?pr=190#step:10:60): in _yield_all_checks
    tags = get_tags(estimator)
/opt/hostedtoolcache/Python/3.9.20/x64/lib/python3.9/site-packages/sklearn/utils/_tags.py:405: in get_tags
    sklearn_tags_provider[klass] = klass.__sklearn_tags__(estimator)  # type: ignore[attr-defined]
/opt/hostedtoolcache/Python/3.9.20/x[64](https://github.com/AdaptiveMotorControlLab/CEBRA/actions/runs/12357735292/job/34486685645?pr=190#step:10:65)/lib/python3.9/site-packages/sklearn/base.py:859: in __sklearn_tags__
    tags = super().__sklearn_tags__()
E   AttributeError: 'super' object has no attribute '__sklearn_tags__'
```

### Operating System

github actions

### CEBRA version

latest main

### Device type

cpu

### Steps To Reproduce

_No response_

### Relevant log output

_No response_

### Anything else?

This issue seems related to the following change in the sklearn API:  https://github.com/scikit-learn/scikit-learn/pull/29677

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/daptiveMotorControlLab/CEBRA/blob/main/CODE_OF_CONDUCT.md)

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -19,10 +19,17 @@ jobs:
         # as well as selected previous versions on
         # https://pytorch.org/get-started/previous-versions/
         torch-version: ["2.2.2", "2.4.0"]
+        sklearn-version: ["latest"]
         include:
           - os: windows-latest
             torch-version: 2.4.0
             python-version: "3.10"
+            sklearn-version: "latest"
+        include:
+          - os: ubuntu-latest 
+            torch-version: 2.4.0
+            python-version: "3.10"
+            sklearn-version: "legacy"
 
     runs-on: ${{ matrix.os }}
 
@@ -32,7 +39,7 @@ jobs:
         uses: actions/cache@v3
         with:
           path: ~/.cache/pip
-          key: pip-os_${{ runner.os }}-python_${{ matrix.python-version }}-torch_${{ matrix.torch-version }}
+          key: pip-os_${{ runner.os }}-python_${{ matrix.python-version }}-torch_${{ matrix.torch-version }}-sklearn_${{ matrix.sklearn-version }}
 
       - name: Checkout code
         uses: actions/checkout@v2
@@ -47,6 +54,11 @@ jobs:
           python -m pip install --upgrade pip setuptools wheel
           python -m pip install torch==${{ matrix.torch-version }} --extra-index-url https://download.pytorch.org/whl/cpu
           pip install '.[dev,datasets,integrations]'
+
+      - name: Check sklearn legacy version 
+        if: matrix.sklearn-version == 'legacy'
+        run: |
+          pip install scikit-learn==1.4.2 '.[dev,datasets,integrations]'
 
       - name: Run the formatter
         run: |

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -25,7 +25,6 @@ jobs:
             torch-version: 2.4.0
             python-version: "3.10"
             sklearn-version: "latest"
-        include:
           - os: ubuntu-latest 
             torch-version: 2.4.0
             python-version: "3.10"

--- a/cebra/integrations/sklearn/cebra.py
+++ b/cebra/integrations/sklearn/cebra.py
@@ -371,7 +371,7 @@ def _load_cebra_with_sklearn_backend(cebra_info: Dict) -> "CEBRA":
     return cebra_
 
 
-class CEBRA(BaseEstimator, TransformerMixin):
+class CEBRA(TransformerMixin, BaseEstimator):
     """CEBRA model defined as part of a ``scikit-learn``-like API.
 
     Attributes:

--- a/cebra/integrations/sklearn/cebra.py
+++ b/cebra/integrations/sklearn/cebra.py
@@ -30,8 +30,10 @@ import numpy.typing as npt
 import pkg_resources
 import sklearn.utils.validation as sklearn_utils_validation
 import torch
+import sklearn
 from sklearn.base import BaseEstimator
 from sklearn.base import TransformerMixin
+from sklearn.utils.metaestimators import available_if
 from torch import nn
 
 import cebra.data
@@ -41,6 +43,11 @@ import cebra.integrations.sklearn.utils as sklearn_utils
 import cebra.models
 import cebra.solver
 
+def check_version(estimator):
+    # NOTE(stes): required as a check for the old way of specifying tags
+    # https://github.com/scikit-learn/scikit-learn/pull/29677#issuecomment-2334229165
+    from packaging import version
+    return version.parse(sklearn.__version__) < version.parse("1.6.dev")
 
 def _init_loader(
     is_cont: bool,
@@ -1294,6 +1301,15 @@ class CEBRA(BaseEstimator, TransformerMixin):
                  callback_frequency=callback_frequency)
         return self.transform(X)
 
+    def __sklearn_tags__(self):
+        # NOTE(stes): from 1.6.dev, this is the new way to specify tags
+        # https://scikit-learn.org/dev/developers/develop.html
+        # https://github.com/scikit-learn/scikit-learn/pull/29677#issuecomment-2334229165
+        tags = super().__sklearn_tags__()
+        tags.non_deterministic = True
+        return tags
+
+    @available_if(check_version)
     def _more_tags(self):
         # NOTE(stes): This tag is needed as seeding is not fully implemented in the
         # current version of CEBRA.

